pkg/asset/agent/manifests/nmstateconfig.go

```diff
@@ -242,41 +243,68 @@ func (n *NMStateConfig) validateNMStateLabels() field.ErrorList {
 	return allErrs
 }
 
-func getFirstIP(nmStateConfig *nmStateConfig) string {
+func getFirstIP(nmstateRaw []byte) (string, error) {
+	var nmStateConfig nmStateConfig
+	err := yaml.Unmarshal(nmstateRaw, &nmStateConfig)
+	if err != nil {
+		return "", fmt.Errorf("error unmarshalling NMStateConfig: %w", err)
+	}
+
 	for _, intf := range nmStateConfig.Interfaces {
 		for _, addr4 := range intf.IPV4.Address {
 			if addr4.IP != "" {
-				return addr4.IP
+				return addr4.IP, nil
 			}
 		}
 		for _, addr6 := range intf.IPV6.Address {
 			if addr6.IP != "" {
-				return addr6.IP
+				return addr6.IP, nil
 			}
 		}
 	}
-	return ""
+
+	return "", nil
 }
 
-// GetNodeZeroIP retrieves the first IP from the user provided NMStateConfigs to set as the node0 IP
-func GetNodeZeroIP(nmStateConfigs []*aiv1beta1.NMStateConfig) (string, error) {
-	for i := range nmStateConfigs {
-		var nmStateConfig nmStateConfig
-		err := yaml.Unmarshal(nmStateConfigs[i].Spec.NetConfig.Raw, &nmStateConfig)
+// GetNodeZeroIP retrieves the first IP to be set as the node0 IP.
+// The method prioritizes the search by trying to scan first the NMState configs defined
+// in the agent-config hosts - so that it would be possible to skip the worker nodes - and then
+// the NMStateConfig.
+func GetNodeZeroIP(agentConfig *agenttypes.Config, nmStateConfigs []*aiv1beta1.NMStateConfig) (string, error) {
+	rawConfigs := []aiv1beta1.RawNetConfig{}
+
+	// Select first the configs from the hosts, if defined
+	if agentConfig != nil {
+		for _, host := range agentConfig.Hosts {
+			// Skip worker hosts
+			if host.Role == "worker" {
+				continue
+			}
+			rawConfigs = append(rawConfigs, host.NetworkConfig.Raw)
+		}
+	}
+
+	// Then pick up the ones from nmStateConfigs, if defined
```

Comment:
Although probably not a problem, it seems that you should have a check here to ignore the settings in nmstateconfig.yaml if agent-config.yaml had the nmstate configuration, in other words `if rawConfigs > 0`.

Reply:
The tricky thing here is that nmstateconfig is both an alternative input to agent-config _and_ a sequential step in a series of transformations starting with agent-config that the user can edit at any time.

In an ideal world we would match each nmstateconfig back to it's corresponding host in agent-config (if there is one) and respect any role settings provided there. Otherwise we could get inconsistent behaviour where the user has edited the nmstateconfig but those changes aren't reflected in the final IP selection. In practice that would require a _lot_ of very gnarly logic.

I think a good compromise would be to treat the nmstateconfigs strictly as an alternative - i.e. only add them to the list when `agentConfig == nil || len(agentConfig.Hosts) == 0`.

There's still potential for inconsistency if the user does crazy stuff, but when combined with the change I suggested above we'd avoid selecting an IP from a host explicitly marked as a worker, without introducing any regressions.